components/raftstore-v2/src/operation/command/admin/flashback.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_traits::{KvEngine, RaftEngine, RaftLogBatch};
+use fail::fail_point;
+use kvproto::{
+    raft_cmdpb::{AdminCmdType, AdminRequest, AdminResponse, RaftCmdRequest},
+    raft_serverpb::RegionLocalState,
+};
+use protobuf::Message;
+use raftstore::{coprocessor::RegionChangeReason, store::metrics::PEER_ADMIN_CMD_COUNTER, Result};
+
+use super::AdminCmdResult;
+use crate::{
+    batch::StoreContext,
+    fsm::ApplyResReporter,
+    raft::{Apply, Peer},
+};
+
+#[derive(Debug)]
+pub struct FlashbackResult {
+    index: u64,
+    region_state: RegionLocalState,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn propose_flashback<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let data = req.write_to_bytes().unwrap();
+        self.propose(store_ctx, data)
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    pub fn apply_flashback(
+        &mut self,
+        index: u64,
+        req: &AdminRequest,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        let is_in_flashback = req.get_cmd_type() == AdminCmdType::PrepareFlashback;
+        // Modify the region meta in memory.
+        let region = self.region_state_mut().mut_region();
+        region.set_is_in_flashback(is_in_flashback);
+        if is_in_flashback {
+            region.set_flashback_start_ts(req.get_prepare_flashback().get_start_ts());
+        } else {
+            region.clear_flashback_start_ts();
+        }
+
+        match req.get_cmd_type() {
+            AdminCmdType::PrepareFlashback => {
+                PEER_ADMIN_CMD_COUNTER.prepare_flashback.success.inc();
+            }
+            AdminCmdType::FinishFlashback => {
+                PEER_ADMIN_CMD_COUNTER.finish_flashback.success.inc();
+            }
+            _ => unreachable!(),
+        }
```

Comment:
The code can be polished by using match req.get_cmd_type() to organizing the whole logic. 
Now it's using req.get_cmd_type() == AdminCmdType::PrepareFlashback initially and then have a match later.

Reply:
```suggestion
        let region = self.region_state_mut().mut_region();
         match req.get_cmd_type() {
            AdminCmdType::PrepareFlashback => {
                PEER_ADMIN_CMD_COUNTER.prepare_flashback.success.inc();
                region.set_is_in_flashback(true);
                region.set_flashback_start_ts(req.get_prepare_flashback().get_start_ts());
            }
            AdminCmdType::FinishFlashback => {
                PEER_ADMIN_CMD_COUNTER.finish_flashback.success.inc();
                region.set_is_in_flashback(false);
                region.clear_flashback_start_ts();
            }
            _ => unreachable!(),
        }
```